### PR TITLE
[mta] Backward of unary foreach functions

### PR DESCRIPTION
**Pull request by crcrpar**
as per title, this PR defines backward of those.

This doesn't implement forward-mode automatic differentiation as [the current codegen](https://github.com/pytorch/pytorch/blob/a747326423ed4731996769e3b8eb73eecbdee2d4/tools/autograd/gen_variable_type.py#L1513) doesn't seem to handle `ArrayRef<Tensor>`.

Rel:
- https://github.com/pytorch/pytorch/issues/53796
- https://github.com/pytorch/pytorch/issues/58833

cc @mcarilli @ngimel

### CODE REVIEW

**Comment by crcrpar** [comment]
regarding inplace ops, I locally confirmed the version seems to be bumped appropriately
```
In [8]: tensors = [torch.randn(3).requires_grad_() for _ in range(2)]
In [9]: print([t._version for t in tensors])
[0, 0]
In [10]: torch._foreach_exp_(tensors)
In [11]: print([t._version for t in tensors])
[1, 1]
```

**Comment by ngimel** [comment]
So the backward of foreach will always fall back to slow version, and codegen would make sure that we properly error out if versions don't match? That looks fine, but I'll let @alband take a look.

**Comment by albanD** [comment]
> regarding inplace ops, I locally confirmed the version seems to be bumped appropriately

I think this is only because you're on CPU: the slow code calls the single-Tensor op which properly bumps the version counter.
Running the same thing on CUDA won't work I think.

File: torch/_torch_docs.py [hunk 1]
```diff
@@ -13897,3 +13897,82 @@ def merge_dicts(*dicts):
...
+Apply :func:`torch.{}` to each Tensor of the input list.
+
+If the input tensors meet a certain condition, for example, they have the same dtype,
+are hosted on the same cuda device, and contiguous (including channels_last and channels_last_3d),
+then the operation done with fewer cuda kernel calls.
+
+.. note::
+    Only CPU and CUDA are supported. Forward-mode AD is not supported.
+
+Args:
```
Comment (albanD) [inline torch/_torch_docs.py:13947]: These args are not correct for functions that take more arguments (like round). Should we just skip it altogether here and just recommend to read the original doc where all Tensor input/output are replaced with list of Tensors?

File: torch/_torch_docs.py [hunk 2]
```diff
@@ -13897,3 +13897,82 @@ def merge_dicts(*dicts):
...
+    self (list of Tensors): Input list of Tensors. Each Tensor can have an arbitrary shape, dtype, device, and strides.
```
Comment (albanD) [inline torch/_torch_docs.py:13948]: It works for difference dtype and device? Why do we collect grads to zero out by device/dtype here then: https://github.com/pytorch/pytorch/blob/274d3b24c3f4b6dc8964ee6642e909ce4b992112/torch/optim/optimizer.py#L394 ?

File: torch/_torch_docs.py [hunk 3]
```diff
@@ -13897,3 +13897,82 @@ def merge_dicts(*dicts):
...
+
+Args:
+    self (list of Tensors): Input list of Tensors.
+    """.format(
+            unary_inplace_foreach_func_name, unary_foreach_func_name
+        ),
+    )
+
+add_docstr(
+    torch._foreach_zero_,
```
Comment (albanD) [inline torch/_torch_docs.py:13969]: This one doesn't need to be special cased? torch.zero_ does exist.

**Review by albanD — COMMENTED** [review]
This is a lot of code duplication.
Should we just have the codegen take care of it? The same way we re-use the `foo` formula for `foo_` when no explicit formula for `foo_` is provided. We could codegen a slow `_foreach_foo` formula when it is not provided. WDYT?

File: torch/_torch_docs.py (hunk 2 above)
Comment (crcrpar) [reply torch/_torch_docs.py]: To reduce the number of CUDA kernels launched as possible, I guess.

File: torch/_torch_docs.py (hunk 3 above)
Comment (crcrpar) [reply torch/_torch_docs.py]: as `_foreach_zero` the outplace one doesn't exist.

File: torch/_torch_docs.py (hunk 1 above)
Comment (crcrpar) [reply torch/_torch_docs.py]: That sounds right but I'm not aware of any unary foreach function taking optional arguments, let me see

File: torch/_torch_docs.py (hunk 2 above)
Comment (albanD) [reply torch/_torch_docs.py]: I'm not sure to follow? How does the two have different number of kernel launched?

File: torch/_torch_docs.py (hunk 3 above)
Comment (albanD) [reply torch/_torch_docs.py:14044]: We can special case it above then to only generate these docs if the corresponding `torch._foreach_*` op exists?

File: torch/_torch_docs.py (hunk 2 above)
Comment (crcrpar) [reply torch/_torch_docs.py]: If we pass a list of tensors of different dtypes and/or devices to a foreach function, it'll call the corresponding aten native function the number of Tensors times, i.e. len(self) cuda kernels. With the grouping, it could reduce the number to `len(per_device_and_dtype_grads)`, assuming all the tensors are a CUDA tensor

File: torch/_torch_docs.py (hunk 3 above)
Comment (crcrpar) [reply torch/_torch_docs.py:14044]: I don't understand. I handled torch._foreach_zero_ differently because there's no `torch._foreach_zero`. The for-loop above is to generate the docs of `torch._foreach_foo` and `torch._foreach_foo_`.

What would you recommend? Defining `torch._foreach_zero` for the parity with the other unary foreach functions?

File: torch/_torch_docs.py (hunk 2 above)
Comment (ngimel) [reply torch/_torch_docs.py]: @alband _for_each ops support args on the different device and of the different dtypes, but they'll fall back to slow implementation. If, instead of just calling `for_each` we pre-sort the args, we can call `for_each` with a subset of tensors that'll go to the fast path. Really this sorting should be done in for_each itself though, and not in the optimizer

Comment (albanD) [reply torch/_torch_docs.py]: Ok!
Should we really advertise this then as it will not do what you expect these functions to do.

File: torch/_torch_docs.py (hunk 3 above)
Comment (albanD) [reply torch/_torch_docs.py:14044]: If the for loop above, you could have a guard for each definition: `if hasattr(torch, "_foreach_foo")` then add the doc for it.
That will remove the need to special case zero (and work for others as well in the future).

File: docs/source/torch.rst [hunk 4]
```diff
@@ -592,6 +592,73 @@ BLAS and LAPACK Operations
     triangular_solve
     vdot
 
+Foreach Operations
+~~~~~~~~~~~~~~~~~~
+
+.. warning::
+    This API is in beta and subject to future changes.
```
Comment (albanD) [inline docs/source/torch.rst:599]: nit: beta or prototype?

File: tools/autograd/derivatives.yaml [hunk 5]
```diff
@@ -2927,3 +2927,89 @@
 - name: _reshape_copy(Tensor self, SymInt[] size) -> Tensor
   self: grad.reshape_symint(self.sym_sizes())
   result: auto_linear
+
+# TODO(crcrpar): Consider defining fast path with the help of `multi_tensor_apply`.
+# rel: `DERIVATIVE_SINGLE_FOREACH` CodeTemplate.
+- name: _foreach_exp(Tensor[] self) -> Tensor[]
+  self: grads[i] * result[i].conj()
```
Comment (albanD) [inline tools/autograd/derivatives.yaml:2934]: This formula is a duplicate of the `exp()` function above.
I was thinking that we could extend https://github.com/pytorch/pytorch/blob/f7939b21e1f09521857515f913cd03d13ea2287b/torchgen/api/autograd.py#L325-L360 to re-use that existing formula and put that in the template you already have.

In particular you can redefine the `grad`, `result`, etc that is used in the original formula within the for loop in the template and just dump the formula in there.
WDYT?

Comment (albanD) [reply tools/autograd/derivatives.yaml:2934]: No you won't need to modify the native functions in any way.
This function is just updating the `formula` field everything else is already properly set.

You just need to generate a good `formula` string for it.

Comment (crcrpar) [reply tools/autograd/derivatives.yaml:2934]: I understand it's cleaner by handling things in codegen but
- We could implement a fast path for backward as well as forward.
- with that this PR will cover the whole foreach functions e.g. binary with `Scalar` and `Scalar[]`

Comment (albanD) [reply tools/autograd/derivatives.yaml:2934]: > We could implement a fast path for backward as well as forward, which would not be compatible with that way of defining backward

You can, this logic that generates the formula kicks in only if you didn't provide a formula already. That's why some inplace functions have formulas and others don't.

> with that this PR will cover the whole foreach functions e.g. binary with Scalar and Scalar[]

Not sure what you mean by that?

**Comment by crcrpar** [comment]
Added a codegen and a version bump

File: aten/src/ATen/native/cuda/ForeachFunctors.cuh [hunk 6]
```diff
@@ -7,6 +7,16 @@ namespace at { namespace native {
 
 namespace {
 
+// TODO(crcrpar): Handle version bump in codegen.
+// rel: https://github.com/pytorch/pytorch/blob/9cf84347767c8abb8feba18a9a1baba321eeb8b9/tools/autograd/gen_inplace_or_view_type.py#L481-L482
+inline void maybe_increment_version(TensorList tensors) {
+  for (const auto & t : tensors) {
+    if (t.requires_grad() && at::GradMode::is_enabled()) {
```
Comment (albanD) [inline aten/src/ATen/native/cuda/ForeachFunctors.cuh:14]: This check is not ok. even things that don't require gradients might get saved for backward and thus we need to always bump their version counter.

File: aten/src/ATen/native/cuda/ForeachUnaryOp.cu [hunk 7]
```diff
@@ -73,6 +73,7 @@ template <typename scalar_t, template<class> class Op> void foreach_unary_op_(Te
                                          /* r_args_depth */ 1,
                                          /* res_arg_index */ 0>(),
                           Op<opmath_t>());
+    maybe_increment_version(tensors);
```
Comment (albanD) [inline aten/src/ATen/native/cuda/ForeachUnaryOp.cu:76]: Do I misremember our discussion where you showed me an example where the version was bumped properly? Why does this need to be added?

**Review by albanD — COMMENTED** [review]
Thanks for the update!
All the codegen part looks good! Only questions about the InplaceOrView key side.

File: aten/src/ATen/native/cuda/ForeachUnaryOp.cu (hunk 7 above)
Comment (crcrpar) [reply aten/src/ATen/native/cuda/ForeachUnaryOp.cu:76]: That happened when inputs tensors are a CPU tensor not when fast path is chosen.
Even the functions registered to native_functions.yaml with CUDA key could go into the slow path which just calls aten native function. Therefore I decided to put this manual version bump here

Comment (albanD) [reply aten/src/ATen/native/cuda/ForeachUnaryOp.cu:76]: Ho that's right!
I'm still not sure why the InplaceOrView kernel doesn't get generated automatically to do that already? But we can look into that later.

File: aten/src/ATen/native/cuda/ForeachFunctors.cuh [hunk 8]
```diff
@@ -7,6 +7,16 @@ namespace at { namespace native {
 
 namespace {
 
+// TODO(crcrpar): Handle version bump in codegen.
+// rel: https://github.com/pytorch/pytorch/blob/9cf84347767c8abb8feba18a9a1baba321eeb8b9/tools/autograd/gen_inplace_or_view_type.py#L481-L482
+inline void maybe_increment_version(TensorList tensors) {
+  for (const auto & t : tensors) {
+    if (at::GradMode::is_enabled()) {
```
Comment (albanD) [inline aten/src/ATen/native/cuda/ForeachFunctors.cuh:14]: Ho sorry my comment wasn't clear enough.
The whole if was not ok. This is not file either. Inplace changes in no_grad mode must also be tracked.
In short, you must always bump the version, whatever the state of autograd.

File: aten/src/ATen/native/cuda/ForeachUnaryOp.cu (hunk 7 above)
Comment (crcrpar) [reply aten/src/ATen/native/cuda/ForeachUnaryOp.cu:76]: I thought with https://github.com/pytorch/pytorch/blob/b33d9e2c87f9323e56672fe862fe1de704346a37/torchgen/model.py#L1388-L1405, inplace functions that modify (and return) multiple Tensors don't get caught by https://github.com/pytorch/pytorch/blob/9cf84347767c8abb8feba18a9a1baba321eeb8b9/tools/autograd/gen_inplace_or_view_type.py#L481-L482

**Review by albanD — APPROVED** [review]
Looks good!
Sorry for all the back and forth, this is definitely a challenging one...

**Comment by crcrpar** [comment]
@pytorchbot merge

(knowing this won’t work this time but want to trigger more jobs)

**Comment by albanD** [comment]
You have a `@pytorchbot label` if you need to add label (like the `ciflow/trunk` one) but you are not allowed.
